yt/data_objects/time_series.py

```diff
@@ -435,6 +436,172 @@ def particle_trajectories(
         )
 
+    def _get_by_attribute(
+        self,
+        attribute: str,
+        value: Union[unyt_quantity, tuple[float, Union[Unit, str]]],
+        tolerance: Union[None, unyt_quantity, tuple[float, Union[Unit, str]]] = None,
+        prefer: Literal["nearest", "smaller", "larger"] = "nearest",
+    ) -> "Dataset":
+        r"""
+        Get a dataset at or near to a given value.
+
+        Parameters
+        ----------
+        attribute : str
+            The key by which to retrieve an output, usually 'current_time' or
+            'current_redshift'. The key must be an attribute of the dataset
+            and monotonic.
+        value : unyt_quantity or (value, unit)
+            The value to search for.
+        tolerance : unyt_quantity or (value, unit), optional
+            If not None, do not return a dataset unless the value is
+            within the tolerance value. If None, simply return the
+            nearest dataset.
+            Default: None.
+        prefer : str
+            The side of the value to return. Can be 'nearest', 'smaller' or 'larger'.
+            Default: 'nearest'.
+        """
+
+        if prefer not in ("nearest", "smaller", "larger"):
+            raise ValueError(
+                f"Side must be 'nearest', 'smaller' or 'larger', got {prefer}."
+            )
+
+        # Use a binary search to find the closest value
+        iL = 0
+        iR = len(self._pre_outputs) - 1
+
+        if iL == iR:
+            ds = self[0]
+            if (
+                tolerance is not None
+                and abs(getattr(ds, attribute) - value) > tolerance
+            ):
+                raise ValueError(
+                    f"No dataset found with {attribute} within {tolerance} of {value}."
+                )
+            return ds
+
+        # Check signedness
+        dsL = self[iL]
+        dsR = self[iR]
+        vL = getattr(dsL, attribute)
+        vR = getattr(dsR, attribute)
+
+        if vL < vR:
+            sign = 1
+        elif vL > vR:
+            sign = -1
+        else:
+            raise ValueError(
+                f"{dsL} and {dsR} have both {attribute}={vL}, cannot perform search."
+            )
+
+        if isinstance(value, tuple):
+            value = dsL.quan(*value)
+        if isinstance(tolerance, tuple):
+            tolerance = dsL.quan(*tolerance)
+
+        # Short-circuit if value is out-of-range
+        if not (vL * sign < value * sign < vR * sign):
+            iL = iR = 0
+
+        while iR - iL > 1:
+            iM = (iR + iL) // 2
+            dsM = self[iM]
+            vM = getattr(dsM, attribute)
+            if sign * value < sign * vM:
+                iR = iM
+                dsR = dsM
+            elif sign * value > sign * vM:
+                iL = iM
+                dsL = dsM
+            else:  # Exact match
+                dsL = dsR = dsM
+                break
+
+        if prefer == "smaller":
```

Comment:
So I'm not sure if this needs to happen up in the `while` loop or down here... but I think you need to explicitly handle the case where the chosen value is exactly halfway between two timesteps, particularly when a value for `prefer` has been provided. As this is, the larger value will always be returned when the provided value is halfway, regardless of `prefer` (see my suggested change in `test_time_series.py`). If you prefer the current behavior then I think the behavior should be documented in the docstring and maybe even a log message (in which case you can ignore my suggested change to the test).